### DISCUSSION TITLE
Underlying mechanism of extractors

### DISCUSSION THREAD

**Discussion by nick-lehmann** [actix/actix-web#3004] [Q&A] [answered]
Hello 👋🏻

I'm really intrigued by the extractor mechanism. It is fascinating to me that I can simply create functions with arguments of some type that implements `FromRequest` and all of the sudden, the data is available inside the function body. However, I have no clue how this works under the hood.

For a personal project (some kind of data loader and test runner), I would like to implement something similar. I wanted to take the extractor mechanism as a base to implement a my own "magic" conversion mechanism. After almost an hour of studying the `actix-web` source code, I was unable to locate the place where this magic happens. By that I mean the place where, depending on the type of a function argument, some code is generated that calls the [`SomeExtractor::from_request`](https://github.com/actix/actix-web/blob/e68f87f84f3109b31de2f2361d4c06af691d8bfb/actix-web/src/extract.rs#L18) method and passes this information to the function at the correct argument position. May anyone give my a hint on where to find this?

**Comment by robjtede** [comment, marked as answer]
A couple of links you may find enlightening.

- https://youtu.be/7DOYtnCXucw?t=1057
- https://docs.rs/actix-web/latest/actix_web/trait.Handler.html

**Reply by nick-lehmann** [reply]
Thank you very much, that was exactly the guidance I needed 👍🏻
